### PR TITLE
Updated Adaptive Pooling layers

### PR DESCRIPTION
**Pull request by Susmit-A**
Updated Adaptive Pooling layers to be analogous to torch.nn.Adaptive*pooling layers from PyTorch.

Fixes # (issue)
Adaptive Pooling Layers had constraints regarding shapes of inputs. Some of those constraints have been relaxed.
Outputs are now same as those given by adaptive layers provided by PyTorch.

## Type of change

- [* ] Bug fix
- [ ] New Tutorial
- [ ] Updated or additional documentation
- [ *] Additional Testing
- [ ] New Activation and the changes conform to the [activation contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/activations/README.md#contribution-guidelines)
- [ ] New Callback and the changes conform to the [callback contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/callbacks/README.md#contribution-guidelines)
- [ ] New Image addition and the changes conform to the [image op contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/image/README.md#contribution-guidelines)
- [* ] New Layer and the changes conform to the [layer contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/layers/README.md#contribution-guidelines)
- [ ] New Loss and the changes conform to the [loss contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/losses/README.md#contribution-guidelines)
- [ ] New Metric and the changes conform to the [metric contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/metrics/README.md#contribution-guidelines)
- [ ] New Optimizer and the changes conform to the [optimizer contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/optimizers/README.md#contribution-guidelines)
- [ ] New RNN Cell and the changes conform to the [rnn contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/rnn/README.md#contribution-guidelines)
- [ ] New Seq2seq addition and the changes conform to the [seq2seq contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/seq2seq/README.md#contribution-guidelines)
- [ ] New Text addition and the changes conform to the [text op contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/text/README.md#contribution-guidelines)

# Checklist:

- [* ] I've properly [formatted my code according to the guidelines](https://github.com/tensorflow/addons/blob/master/CONTRIBUTING.md#coding-style)
    - [* ] By running Black + Flake8
    - [* ] By running pre-commit hooks
- [*] This PR addresses an already submitted issue for TensorFlow Addons
- [x] I have made corresponding changes to the documentation
- [* ] I have added tests that prove my fix is effective or that my feature works
- [ ] This PR contains modifications to C++ custom-ops

# How Has This Been Tested?

If you're adding a bugfix or new feature please describe the tests that you ran to verify your changes:
*  The new layers were extensively tested against their PyTorch counterparts.

### CODE REVIEW

**Review by WindQAQ — CHANGES_REQUESTED** [review]
Thanks for the PR. To prevent performance regression, it's better to do the following thing

```python
def _adaptive_pooling_1d(inputs, output_size, data_format, reduction_function):
    def divisible_case():
        # Original codes go here
        pass

    def nondivisible_case():
        # This PR goes here
        pass

    input_shape = tf.shape(inputs)
    length = input_shape[1] if data_format == "channels_last" else input_shape[2]
    is_divisible = (length % output_size == 0)

    return tf.keras.backend.switch(is_divisible, divisible_case, nondivisible_case)


class AdaptivePooling1D(tf.keras.layers.Layer):
    def call(self, inputs):
        return _adaptive_pooling_1d(inputs, self.output_size, self.data_format, self.reduction_function)
```

**Comment by Susmit-A** [comment]
> Thanks for the PR. To prevent performance regression, it's better to do the following thing
> 
> ```python
> def _adaptive_pooling_1d(inputs, output_size, data_format, reduction_function):
>     def divisible_case():
>         # Original codes go here
>         pass
> 
>     def nondivisible_case():
>         # This PR goes here
>         pass
> 
>     input_shape = tf.shape(inputs)
>     length = input_shape[1] if data_format == "channels_last" else input_shape[2]
>     is_divisible = (length % output_size == 0)
> 
>     return tf.keras.backend.switch(is_divisible, divisible_case, nondivisible_case)
> 
> 
> class AdaptivePooling1D(tf.keras.layers.Layer):
>     def call(self, inputs):
>         return _adaptive_pooling_1d(inputs, self.output_size, self.data_format, self.reduction_function)
> ```
> 
> For 2d and 3d, can we have the following patterns to reduce the number of reduction and slicing calls?
> 
> ```python
> for i in range(output_size[0]):
>     for j in range(output_size[1]):
>         for k in range(output_size[2]):
>             sliced_inputs = inputs[:, foo1:bar1, foo2:bar2, foo3:bar3, :]
>             pooled = reduction_function(sliced_inputs, axis=[1, 2, 3])
> ```

The former can be done. The latter, however, will cause the code to slow down further, since it would independently address each piece along each axis (O(n^3)). Independent loops can access one whole axis at a time, so the time for 3D inputs would be (O(3n)), as elements in each access can be processed completely in parallel.

File: tensorflow_addons/layers/adaptive_pooling.py [hunk 1]
```diff
@@ -183,24 +233,132 @@ def __init__(
         self.data_format = conv_utils.normalize_data_format(data_format)
         self.reduce_function = reduce_function
         self.output_size = conv_utils.normalize_tuple(output_size, 2, "output_size")
+        self.output_size_x, self.output_size_y = self.output_size
```
Comment (WindQAQ) [inline tensorflow_addons/layers/adaptive_pooling.py:236]: We can either access through `self.output_size[0]` and `self.output_size[1]` or unpack it in `call` so do not set them as attribute.

File: tensorflow_addons/layers/adaptive_pooling.py [hunk 2]
```diff
@@ -183,24 +233,132 @@ def __init__(
...
+                tf.math.ceil(
+                    (
+                        (tf.range(self.output_size_x, dtype=tf.float32) + 1)
+                        * tf.cast(
+                            (tf.shape(inputs)[1] / self.output_size_x), tf.float32
+                        )
+                    )
+                ),
+                tf.int32,
+            )
```
Comment (WindQAQ) [inline tensorflow_addons/layers/adaptive_pooling.py:259]: can we get some utility functions say

```python
def _start_indices(length, output_size):
    return tf.cast(
                (
                    tf.range(output_size, dtype=tf.float32)
                    * tf.cast((length / output_size), tf.float32)
                ),
                tf.int32,
            )


def _end_indices(length, output_size):
    return tf.cast(
                tf.math.ceil(
                    (
                        (tf.range(output_size, dtype=tf.float32) + 1)
                        * tf.cast(
                            (length / output_size), tf.float32
                        )
                    )
                ),
                tf.int32,
            )
```

File: tensorflow_addons/layers/tests/adaptive_pooling_test.py [hunk 3]
```diff
@@ -179,7 +171,7 @@ def test_max_3d():
     ).astype(np.float32)
     output = np.reshape(output, (1, 1, 2, 2, 2))
     test_utils.layer_test(
-        AdaptiveMaxPooling3D,
+        adaptive_pooling.AdaptiveMaxPooling3D,
         kwargs={"output_size": (2, 2, 2), "data_format": "channels_first"},
         input_data=valid_input,
         expected_output=output,
```
Comment (WindQAQ) [inline tensorflow_addons/layers/tests/adaptive_pooling_test.py:177]: Also add some non-divisible cases and numerical tests.

**Comment by WindQAQ** [comment]
> The former can be done. The latter, however, will cause the code to slow down further, since it would independently address each piece along each axis (O(n^3)). Independent loops can access one whole axis at a time, so the time for 3D inputs would be (O(3n)), as elements in each access can be processed completely in parallel.

Great, I am not aware that :-)

**Comment by seanpmorgan** [comment]
Thank you for your contribution. We sincerely apologize for any delay in reviewing, but TensorFlow Addons is transitioning to a minimal maintenance and release mode. New features will not be added to this repository. For more information, please see our public messaging on this decision:
[TensorFlow Addons Wind Down](https://github.com/tensorflow/addons/issues/2807)

Please consider sending feature requests / contributions to other repositories in the TF community with a similar charters to TFA:
[Keras](https://github.com/keras-team/keras)
[Keras-CV](https://github.com/keras-team/keras-cv)
[Keras-NLP](https://github.com/keras-team/keras-nlp)
